storage/google_storage.py

```diff
@@ -3,53 +3,66 @@
 import json
 import os
 
-import google.cloud.storage
-import google.oauth2.service_account
+import google.cloud.storage  # type: ignore
+from google.cloud.storage.bucket import Bucket  # type: ignore
+from google.cloud.storage.blob import Blob  # type: ignore
+import google.oauth2.service_account  # type: ignore
 
-from . import retry
-from .storage import Storage, register_storage_protocol
+from typing import BinaryIO, cast, Optional
+
+from storage import retry
+from storage.storage import Storage, register_storage_protocol
+
+
+class GoogleStorageException(Exception):
+    pass
 
 
 @register_storage_protocol("gs")
 class GoogleStorage(Storage):
-    def _get_bucket(self):
-        credentials_data = json.loads(base64.urlsafe_b64decode(
-            self._parsed_storage_uri.username.encode("utf8")))
+
+    def _get_bucket(self) -> Bucket:
+        username = self._parsed_storage_uri.username
+        credentials_data = json.loads(base64.urlsafe_b64decode(username))
         credentials = google.oauth2.service_account.Credentials.from_service_account_info(
             credentials_data)
         client = google.cloud.storage.client.Client(
             project=credentials_data["project_id"], credentials=credentials)
         return client.get_bucket(self._parsed_storage_uri.hostname)
 
-    def _get_blob(self):
+    def _get_blob(self) -> Blob:
         bucket = self._get_bucket()
-        return bucket.blob(self._parsed_storage_uri.path[1:])
+        blob = bucket.blob(self._parsed_storage_uri.path[1:])
+        if blob is None:
+            raise GoogleStorageException(
+                f"Could not find resource {self._parsed_storage_uri.path}")
+        return blob
 
-    def save_to_filename(self, file_path):
+    def save_to_filename(self, file_path: str) -> None:
         blob = self._get_blob()
         blob.download_to_filename(file_path)
 
-    def save_to_file(self, out_file):
+    def save_to_file(self, out_file: BinaryIO) -> None:
         blob = self._get_blob()
         blob.download_to_file(out_file)
 
-    def load_from_filename(self, file_path):
+    def load_from_filename(self, file_path: str) -> None:
         blob = self._get_blob()
         blob.upload_from_filename(file_path)
 
-    def load_from_file(self, in_file):
+    def load_from_file(self, in_file: BinaryIO) -> None:
         blob = self._get_blob()
         blob.upload_from_file(in_file)
 
-    def delete(self):
+    def delete(self) -> None:
         blob = self._get_blob()
         blob.delete()
 
-    def get_download_url(self, seconds=60, key=None):
+    def get_download_url(self, seconds: int = 60, key: Optional[str] = None) -> str:
         blob = self._get_blob()
-        return blob.generate_signed_url(datetime.timedelta(seconds=seconds))
+        return cast(str, blob.generate_signed_url(datetime.timedelta(seconds=seconds)))
```

Comment:
This is to make the type checker happy

Reply:
To be specific, this is necessary because Google Storage library doesn't have types, and we are ignoring the types on import time, which means the result of any library, method, or attribute lookup is going to be `Any`. Alternatively, we could implement stubs for every single library we use, but that felt out-of-scope, especially since they will probably find their way to `typeshed`, etc. in the near-ish future anyway.